contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,588 @@
+
+  /*
+   (productId, poolAddress) => lastPriceUpdate
+   Last base price update time.
+  */
+  mapping(uint => mapping(address => uint)) lastPriceUpdate;
+
+
+  /* === CONSTANTS ==== */
+
+  uint public REWARD_BPS = 5000;
+  uint public constant PERCENTAGE_CHANGE_PER_DAY_BPS = 100;
+  uint public constant BASIS_PRECISION = 10000;
+  uint public constant STAKE_SPEED_UNIT = 100000e18;
+  uint public constant PRICE_CURVE_EXPONENT = 7;
+  uint public constant MAX_PRICE_PERCENTAGE = 1e20;
+
+  /* === MODIFIERS ==== */
+
+  modifier onlyAdvisoryBoard {
+    uint abRole = uint(IMemberRoles.Role.AdvisoryBoard);
+    require(
+      memberRoles().checkRole(msg.sender, abRole),
+      "Cover: Caller is not an advisory board member"
+    );
+    _;
+  }
+
+  constructor(string memory name_, string memory symbol_) ERC721(name_, symbol_) {
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+  function buyCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external payable override returns (uint /*coverId*/) {
+    require(initialPrices[productId] != 0, "Cover: product not initialized");
+
+    (uint coverId, uint priceInAsset) = _createCover(owner, productId, payoutAsset, amount, period, stakingPools);
+    require(priceInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(priceInAsset, payoutAsset);
+    return coverId;
+  }
+
+  function _createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint coverId, uint premiumInAsset) {
+
+    // convert to NXM amount
+    uint amountLeftToCoverInNXM = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(payoutAsset));
+    activeCoverAmountInNXM[productId] += uint96(amountLeftToCoverInNXM);
+
+    uint totalPremiumInNXM = 0;
+
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountLeftToCoverInNXM == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint premiumInNXM) = buyCoverFromPool(stakingPool, productId, amountLeftToCoverInNXM, period);
+      amountLeftToCoverInNXM -= coveredAmount;
+      totalPremiumInNXM += premiumInNXM;
+      stakingPoolsForCover[covers.length].push(
+        StakingPool(address(stakingPool), uint96(coveredAmount), uint96(premiumInNXM))
+      );
+    }
+    require(amountLeftToCoverInNXM == 0, "Not enough available capacity");
+
+    premiumInAsset = totalPremiumInNXM * pool().getTokenPrice(pool().assets(payoutAsset)) / 1e18;
+
+    covers.push(Cover(
+        productId,
+        payoutAsset,
+        uint96(amount),
+        uint32(block.timestamp + 1),
+        uint32(period),
+        uint96(premiumInAsset)
+      ));
+
+    coverId = covers.length - 1;
+    _safeMint(owner, coverId);
+  }
```

Comment:
this function is expected to return: `(uint coverId, uint priceInAsset) = _createCover(...)`, but it doesn't